mplc/multi_partner_learning.py

```diff
@@ -60,594 +53,427 @@ def __init__(self,
-        mb_nb_str = f"Minibatch {str(self.minibatch_index).zfill(2)}/{str(self.minibatch_count - 1).zfill(2)}"
-        partner_id_str = f"Partner id #{partner.id} ({partner_index}/{self.partners_count - 1})"
-        val_acc_str = f"{round(validation_score, 2)}"
-
-        logger.debug(f"{epoch_nb_str} > {mb_nb_str} > {partner_id_str} > val_acc: {val_acc_str}")
-
-    def update_iterative_results(self, partner_index, fit_history):
-        """Update the results arrays with results from the collaboration round"""
-
-        validation_score = fit_history.history["val_accuracy"][0]
-
-        self.scores_last_learning_round[partner_index] = validation_score  # TO DO check if coherent
-
-        # At the end of each mini-batch, for each partner, populate the score matrix per partner
-        self.score_matrix_per_partner[self.epoch_index, self.minibatch_index, partner_index] = validation_score
+class SequentialWithFinalAggLearning(SequentialLearning):
+    def __init__(self,
+                 partners_list,
+                 epoch_count,
+                 minibatch_count,
+                 dataset,
+                 aggregation=UniformAggregator,
+                 is_early_stopping=True,
+                 is_save_data=False,
+                 save_folder="",
+                 init_model_from="random_initialization",
+                 use_saved_weights=False,
+                 ):
+        super(SequentialWithFinalAggLearning, self).__init__(partners_list,
+                                                             epoch_count,
+                                                             minibatch_count,
+                                                             dataset,
+                                                             aggregation,
+                                                             is_early_stopping,
+                                                             is_save_data,
+                                                             save_folder,
+                                                             init_model_from,
+                                                             use_saved_weights)
+        if self.partners_count == 1:
+            raise ValueError('Only one partner is provided. Please use the dedicated SinglePartnerLearning class')
+
+    def fit_epoch(self):
+        # Clear Keras' old models
+        clear_session()
+
+        # Split the train dataset in mini-batches
+        self.split_in_minibatches()
+
+        # Iterate over mini-batches and train
+        for i in range(self.minibatch_count):
+            logger.info(f"(seq-final-agg) Minibatch n°{self.minibatch_index} of epoch n°{self.epoch_index}, "
+                        f"init model with a copy of the global model")
+            self.minibatch_index = i
+            self.fit_minibatch()
+
+        # At the end of each epoch, aggregate the models
+        self.model_weights = self.aggregator.aggregate_model_weights()
```

Comment:
Hmm, small detail that had its importance:

- in seq-with-final-agg, the aggregation is made as I explained last time for having a more reliable end-of-epoch perf eval for early stopping (and final perf test)
- however, initially what we had designed was that the next epoch shouldn't start from the aggregated model. It should start from the final sequential model of the last epoch. So that the learning stays really sequential
- however, thinking about it, I kind of like this intermediate implementation where it starts from the aggregated model. We could keep it like that (although it probably means updating the schema in the documentation)